jl4-lsp/src/LSP/L4/Actions.hs

```diff
@@ -96,6 +103,54 @@ gotoDefinition pos m positionMapping = do
   newRange <- toCurrentRange positionMapping lspRange
   pure (Location (fromNormalizedUri range.moduleUri) newRange)
 
+
+-- ----------------------------------------------------------------------------
+-- Ladder evalApp
+-- ----------------------------------------------------------------------------
+
+evalApp
+  :: MonadIO m
+  => TypeCheckResult
+  -> Ladder.EvalAppRequestParams
+  -> RecentlyVisualised
+  -> EL.Environment
+  -> ExceptT (TResponseError method) m Aeson.Value
+evalApp tcRes evalParams recentViz evalEnv =
+  case Ladder.lookupEvalAppMaker recentViz.vizState evalParams.appExpr of
+    Nothing -> defaultResponseError "No eval app directive maker found" -- TODO: Improve error codehere
+    Just evalAppMaker -> do
+      let evalAppDirective = evalAppMaker evalParams
+      (_, results) <- liftIO $ EL.execEvalModuleWithEnv evalEnv (evalAppDirective `prependToModule` tcRes.module')
```

Comment:
If you don't want to use the rule directly and that's why you decided to use the function here instead, you can pass in a continuation that uses the rule from the callsite of this function and then just use the continuation in place of execEvalModuleWithEnv

Reply:
I'm using the execEval... function because I didn't see an easy way to use the `EvaluateLazy` rule on a new module (that consists of the evalAppDirective prepended to the old). But I do use the GetLazyEvaluationDependencies rule in the handler to get the eval environment and then pass that in to the `evalApp` call.

The EvalLazy rule, for reference:
```
  define shakeRecorder $ \EvaluateLazy uri -> do
    res  <- use_ (AttachCallStack [uri] GetLazyEvaluationDependencies) uri
    let results = snd res
    pure (mkSimpleFileDiagnostic uri . evalLazyResultToDiagnostic <$> results, Just results)
  ```

Reply:
I could also make a new rule with the eval-related logic from the `evalApp` function, if you think that's better.

Reply:
Part of what I'm struggling with here is that it's not clear to me what the underlying motivation / concern is exactly. E.g. is it about caching computation as much as possible (in which case I should probably factor the key logic from evalApp out into a new rule)? Or is it more about not using `execEvalModuleWithEnv` directly? Or is it about having all the rules-related logic in one place?